addons/wiki/static/dragNDrop.js

```diff
@@ -0,0 +1,341 @@
+
+        /**
+         * Sets the values of the input area to be the current values of init.before, init.selection, and init.after
+         *
+         * init.before = everything before cursor/selection
+         *
+         * init.selection = text that is highlighted
+         *
+         * init.after = everything after cursor.selection
+         */
+        var fixupInputArea = function() {
+            state.setChunks(init);
+            state.restore();
+        };
+
+        /**
+         * If the item being dragged is from elsewhere online, html and/or URL will be defined
+         *
+         * html will be the HTML block for the element being dragged
+         *
+         * url will be some sort of url that is hopefully an image url (or an image can be parsed out)
+         *
+         * remoteFileHandler() will attempt to figure this out and react accordingly
+         */
+        var html = event.dataTransfer.getData('text/html');
+        var url = event.dataTransfer.getData('URL');
+        if (!!html || !!url) {
+            remoteFileHandler(html, url, cm, init, fixupInputArea);
+        } else {
+            /**
+             * If event.dataTransfer does not have html or url for the item(s), then try to upload it as a file
+             *
+             * localFileHandler() will deal with all of the error checking/handling for this
+             */
+            var files = event.dataTransfer.files;
+            localFileHandler(files, cm, init, fixupInputArea);
+        }
+        editor.marker.session.removeMarker(editor.marker.id);
+        editor.marker.redraw();
+        editor.marker.active = false;
+    }, true);
+
+    /**
+     * Called if something is dragged over the editor and then dragged back out
+     *
+     * Removes the second cursor
+     */
+    element.addEventListener('dragleave', function(event) {
+        event.preventDefault();
+        event.stopPropagation();
+        editor.marker.session.removeMarker(editor.marker.id);
+        editor.marker.redraw();
+        editor.marker.active = false;
+    });
+};
+
+var imageFolder = 'Wiki images';
+
+var notUploaded = function(response, multiple, cm, init, fixupInputArea, path, file) {
+    var files = multiple ? 'File(s)' : 'File';
+    if (response.status === 403) {
+        $osf.growl('Error', 'File not uploaded. You do not have permission to upload files to' +
+            ' this project.', 'danger');
+        editor.enable();
+    } else if (response.status === 409) {
+        // If you drag something with the same name redraw the link, but tell the user what to do in
+        // case they want to update the image with the same name.
+        $.ajax({
+            url: ctx.waterbutlerURL + 'v1/resources/' + ctx.node.id + '/providers/osfstorage' + path,
+            beforeSend: $osf.setXHRAuthorization,
+            success: function(response) {
```

Comment:
Minor: As per our conventions, use the promise interface (`.then()`) rather than the `success` option.